core/drivers/dra7_rng.c

```diff
@@ -0,0 +1,217 @@
+#define	RNG_SOFT_RESET_REG      0x1FF0
+#  define RNG_SOFT_RESET        BIT(0)
+#define	RNG_IRQ_EOI_REG         0x1FF4
+#define	RNG_IRQSTATUS           0x1FF8
+
+#define RNG_CONTROL_STARTUP_CYCLES_SHIFT        16
+#define RNG_CONTROL_STARTUP_CYCLES_MASK         (0xffff << 16)
+
+#define RNG_CONFIG_MAX_REFIL_CYCLES_SHIFT       16
+#define RNG_CONFIG_MAX_REFIL_CYCLES_MASK        (0xffff << 16)
+#define RNG_CONFIG_MIN_REFIL_CYCLES_SHIFT       0
+#define RNG_CONFIG_MIN_REFIL_CYCLES_MASK        (0xff << 0)
+
+#define RNG_ALARMCNT_ALARM_TH_SHIFT             0x0
+#define RNG_ALARMCNT_ALARM_TH_MASK              (0xff << 0)
+#define RNG_ALARMCNT_SHUTDOWN_TH_SHIFT          16
+#define RNG_ALARMCNT_SHUTDOWN_TH_MASK           (0x1f << 16)
+
+#define RNG_CONTROL_STARTUP_CYCLES              0xff
+#define RNG_CONFIG_MIN_REFIL_CYCLES             0x21
+#define RNG_CONFIG_MAX_REFIL_CYCLES             0x22
+#define RNG_ALARM_THRESHOLD                     0xff
+#define RNG_SHUTDOWN_THRESHOLD                  0x4
+
+#define RNG_FRO_MASK    0xffffff
+
+#define RNG_REG_SIZE    0x2000
+
+register_phys_mem(MEM_AREA_IO_SEC, RNG_BASE, RNG_REG_SIZE);
+
+static struct mutex rng_mutex = MUTEX_INITIALIZER;
+
+static vaddr_t rng_base(void)
+{
+	static void *va __early_bss;
+
+	if (cpu_mmu_enabled()) {
+		if (!va)
+			va = phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC);
+		return (vaddr_t)va;
+	}
+
+	return RNG_BASE;
+}
+
+uint8_t hw_get_random_byte(void)
+{
+	static vaddr_t rng;
+	static int pos;
+	static union {
+		uint64_t val;
+		uint8_t byte[8];
+	} random;
+	uint8_t ret;
+
+	mutex_lock(&rng_mutex);
+
+	if (!rng)
+		rng = rng_base();
+
+	if (!pos) {
+		/* Is the result ready (available)? */
+		while (!(read32(rng + RNG_STATUS) & RNG_READY)) {
+			/* Is the shutdown threshold reached? */
+			if (read32(rng + RNG_STATUS) & SHUTDOWN_OFLO) {
+				uint32_t alarm = read32(rng + RNG_ALARMSTOP);
+				uint32_t tuning = read32(rng + RNG_FRODETUNE);
+				/* Clear the alarm events */
+				write32(0x0, rng + RNG_ALARMMASK);
+				write32(0x0, rng + RNG_ALARMSTOP);
+				/* De-tune offending FROs */
+				write32(tuning ^ alarm, rng + RNG_FRODETUNE);
+				/* Re-enable the shut down FROs */
+				write32(RNG_FRO_MASK, rng + RNG_FROENABLE);
+				/* Clear the shutdown overflow event */
+				write32(SHUTDOWN_OFLO, rng + RNG_INTACK);
+
+				DMSG("Fixed FRO shutdown\n");
+			}
+		}
+		/* Read random value */
+		random.val = read32(rng + RNG_OUTPUT_L);
+		random.val |= SHIFT_U64(read32(rng + RNG_OUTPUT_H), 32);
+		/* Acknowledge read complete */
+		write32(RNG_READY, rng + RNG_INTACK);
+	}
+
+	ret = random.byte[pos++];
+
+	if (pos == 8)
+		pos = 0;
+
+	mutex_unlock(&rng_mutex);
+
+	return ret;
+}
+
+static TEE_Result dra7_rng_init(void)
+{
+	vaddr_t rng = rng_base();
+	uint32_t val;
+
+	/* Execute a software reset */
+	write32(RNG_SOFT_RESET, rng + RNG_SOFT_RESET_REG);
+
+	/* Wait for the software reset completion by polling */
+	while (read32(rng + RNG_SOFT_RESET_REG) & RNG_SOFT_RESET)
+		;
```

Comment:
I'm not sure what an appropriate timeout would be, I'm not finding a maximum expected reset time in the datasheet.